### PR TITLE
Subnet registry to deploy upgradeable subnet actors

### PR DESCRIPTION
**Pull request by snissn**
Change subnet registry so that subnet actors the registry deploys have the diamondcut and diamondLoupe Factes on them
 
- [x] **Blocker: ENG-29 RegisterSubnetFacet diamond constructor missing Loupe and DiamondCut**
  - [x] add code from test createSubnetActorDiamondWithFaucets to subnet registry function `newSubnetActor`
  - [x] update `SubnetRegistryDiamond.sol` for 2 more facets
    - [x] add to `constructorParams`
      - `bytes4[] subnetActorDiamondCutSelectors;`
      - `bytes4[] subnetActorLoupeSelectors;`
      - `address diamondCutFacet;`
      - `address loupeFacet;`
    - [x] in constructor
      - [x] assert both addresses != 0
    - [x] set state variable "s" with the four values above
    - [x] update `./lib/LibSubnetRegistryStorage.sol`
      - `address SUBNET_ACTOR_DIAMOND_CUT_FACET;`
      - `address SUBNET_ACTOR_LOUPE_FACET;`
      - `bytes4[] subnetActorDiamondCutSelectors;`
      - `bytes4[] subnetActorLoupeSelectors;`

- [x] `deploy-registry.ts` needs to deploy DiamondCut and Loupe facets

 ## Upgrade Code
- [x] Use registry config and after upgrading create a config for each subnet actor that is upgraded in the format `subnet.actor-{address.}json`
- [x]  Update subnet actor upgrade scripts to accept an address for the subnet actor

## Testing plan
1. Deploy a new gateway and registry combo
2. update ~/.ipc/config with latest gateway and registry info
3. create a new subnet using the command `ipc-cli subnet create --parent /r1337  --min-validators 3 --min-validator-stake 1 --bottomup-check-period 30 --permission-mode collateral --supply-source-kind native --from 0xAddress`
4. modify one of the subnet actor facets, such as deleting or adding a function, or editing a function and run this command to perform an upgrade: `ipc/contracts$ make upgrade-sa-diamond SUBNET_ACTOR_ADDRESS=0xed27b8e0f2f382195d1a71d55edccb9acccb6daf`
5. The subnet actor should be upgraded on chain with an output similar to:
```
make upgrade-sa-diamond SUBNET_ACTOR_ADDRESS=0xed27b8e0f2f382195d1a71d55edccb9acccb6daf
./ops/upgrade-sa-diamond.sh auto 0xed27b8e0f2f382195d1a71d55edccb9acccb6daf
[*] Automatically getting chainID for network
[*] Target network Chain ID: 0x539
Generating typings for: 1 artifacts in dir: typechain for target: ethers-v5
Successfully generated 56 typings!
Compiled 1 Solidity file successfully (evm target: paris).

Facet Bytecode Not Found:
---------------------------------
Facet Name: SubnetActorManagerFacet
Libraries:

Address: 0x5172193b895876021684E8b20323Ac0C7467351c


Diamond Facet Upgrade:
-----------------------------------
Diamond Address: 0xed27b8e0f2f382195d1a71d55edccb9acccb6daf
Replacement Facet Name: SubnetActorManagerFacet


Deployment Status:
-------------------------
New replacement facet (SubnetActorManagerFacet) deployed.
```
6. confirm a new file is created `subnet.actor-0xed27b8e0f2f382195d1a71d55edccb9acccb6daf.json` and archive that information for the future because subsequent upgrades from this make file will expect that to be there

### CODE REVIEW

File: contracts/Makefile [hunk 1]
```diff
@@ -20,7 +20,7 @@ upgrade-gw-diamond:
 	./ops/upgrade-gw-diamond.sh $(NETWORK)
 
 upgrade-sa-diamond:
-	./ops/upgrade-sa-diamond.sh $(NETWORK)
+	./ops/upgrade-sa-diamond.sh $(NETWORK) $(SUBNET_ACTOR_ADDRESS)
 
 upgrade-sr-diamond:
```
Comment (cryptoAtwill) [inline contracts/Makefile:25]: Hi @snissn, just a quick question for upgrading `subnet-registry`, if the subnet actor code has updated, and an we want the registry to deploy new version of `SubnetActor`, after an upgrade of SubnetRegistry, will the registry start deploying the newer version of the Subnet Actor?

Comment (snissn) [reply contracts/Makefile:25]: Hey will! The answer should be yes. Using the upgrade code to upgrade `src/subnetregistry/RegisterSubnetFacet.sol` in the registry should change the deploy path for new subnets.

File: contracts/src/subnetregistry/RegisterSubnetFacet.sol [hunk 2]
```diff
@@ -57,6 +57,18 @@ contract RegisterSubnetFacet is ReentrancyGuard {
             functionSelectors: s.subnetActorPauserSelectors
         });
 
+        diamondCut[5] = IDiamond.FacetCut({
+            facetAddress: s.SUBNET_ACTOR_DIAMOND_CUT_FACET,
```
Comment (crystalbit) [inline contracts/src/subnetregistry/RegisterSubnetFacet.sol:61]: Is it reasonable to add OwnershipFacet here? We need owner transferring

Comment (snissn) [reply contracts/src/subnetregistry/RegisterSubnetFacet.sol:61]: perfect :heavy_check_mark:

File: fendermint/vm/actor_interface/src/ipc.rs [hunk 3]
```diff
@@ -64,6 +64,18 @@ lazy_static! {
...
+                            abi: ia::diamond_loupe_facet::DIAMONDLOUPEFACET_ABI.to_owned(),
+                        },
+                        EthFacet {
+                            name: "DiamondCutFacet",
+                            abi: ia::diamond_cut_facet::DIAMONDCUTFACET_ABI.to_owned(),
+                        },
+                        EthFacet {
+                            name: "OwnershipFacet",
+                            abi: ia::ownership_facet::OWNERSHIPFACET_ABI.to_owned(),
+                        },
```
Comment (raulk) [inline fendermint/vm/actor_interface/src/ipc.rs:78]: We were missing these facets from the gateway. Although the owner address will be f01 (Init actor), as that's the actor that creates the gateway and the registry during genesis:

https://github.com/consensus-shipyard/ipc/blob/f9bda9e83dcd6c7175a25df604f16aec40ffa973/fendermint/vm/interpreter/src/fvm/state/genesis.rs#L434

Which means that we anyway can't upgrade the gateway and registry actors inside a subnet without a hard fork.

cc @fridrik01 @cryptoAtwill

**Review by raulk — APPROVED** [review]

**Pull request merged**